torch/distributed/rpc/rref_proxy.py

```diff
@@ -15,24 +15,82 @@ def _local_invoke_async_execution(rref, func_name, args, kwargs):
-        rref_type, torch._C.ScriptModule
-    )
-    if not bypass_type:
-        func = getattr(rref_type, func_name)
-        if hasattr(func, "_wrapped_async_rpc_function"):
-            _invoke_func = _local_invoke_async_execution
-
-    return rpc_api(
-        rref.owner(),
-        _invoke_func,
-        args=(rref, func_name, args, kwargs),
-        timeout=timeout
-    )
+    rref_type = rref._get_type(timeout=timeout, blocking=False)
+    # Controls whether we are running in a callback or inline in invoke_on_owner
+    # to appropriately handle errors.
+    block_on_type = rpc_api in [
+        torch.distributed.rpc.rpc_sync, torch.distributed.rpc.remote
+    ]
+
+    # Helper function to allow invoke to be run as a chained callback instead of
+    # inline.
+
+    def invoke_on_owner(rref_type):
+        _invoke_func = _local_invoke
+        if not block_on_type:
+            # Running as a callback, need to propagate error due to
+            # https://github.com/pytorch/pytorch/issues/52132
+            try:
+                rref_type = rref_type.wait()
+            except BaseException as err:
+                # Future corresponding to rref type had an error, propagate
+                # error instead of issuing RPC
+                ret = torch.futures.Future()  # type: ignore
+                ret.set_exception(err)
+                return ret
+        else:
+            rref_type = rref_type.wait()
+
+        # Bypass ScriptModules when checking for async function attribute.
+        bypass_type = issubclass(rref_type, torch.jit.ScriptModule) or issubclass(
+            rref_type, torch._C.ScriptModule
+        )
+        if not bypass_type:
+            if not block_on_type:
+                # Running as a callback, need to propagate error due to
+                # https://github.com/pytorch/pytorch/issues/52132
+                try:
+                    func = getattr(rref_type, func_name)
+                except BaseException as err:
+                    ret = torch.futures.Future()  # type: ignore
+                    ret.set_exception(err)
+                    return ret
+            else:
+                func = getattr(rref_type, func_name)
+            if hasattr(func, "_wrapped_async_rpc_function"):
+                _invoke_func = _local_invoke_async_execution
+
+        return rpc_api(
+            rref.owner(),
+            _invoke_func,
+            args=(rref, func_name, args, kwargs),
+            timeout=timeout
+        )
+
+    # If we are not running rpc_async, invoke the RPC API inline
+    if block_on_type:
+        return invoke_on_owner(rref_type)
+    else:
+        # Create a future whose result is the result of the above rpc_api()
+        # future. We don't simply return rref_type.then(cb) since the user would
+        # have to call wait() twice to access the future result.
+        ret_fut = torch.futures.Future()  # type: ignore
```

Comment:
To make `remote` async as well, you could create an RRef in Python and then have a private RPC `remote` API that accepts a python created user rref instead of creating one internally. Then you can just return this RRef to the user and still have async behavior for `remote`.